pkg/sbom/diff_nodelist.go

```diff
@@ -0,0 +1,256 @@
+package sbom
+
+import (
+	"regexp"
+	"sort"
+	"strings"
+)
+
+// NodeListDiff represents the difference between two lists NodeLists.
+type NodeListDiff struct {
+	NodesDiff         NodeListDiffNodes
+	EdgesDiff         NodeListDiffEdges
+	RootElmementsDiff NodeListRootElementDiff
+}
+
+func (nlDiff *NodeListDiff) AddedNodeList() *NodeList {
+	var addedNodeList NodeList
+	for _, addedNode := range nlDiff.NodesDiff.Added {
+		addedNodeList.AddNode(addedNode)
+	}
+	for _, addedEdge := range nlDiff.EdgesDiff.Added {
+		addedNodeList.AddEdge(addedEdge)
+	}
+	addedNodeList.RootElements = append(addedNodeList.RootElements, nlDiff.RootElmementsDiff.Added...)
+
+	for _, diffNodeAdded := range nlDiff.NodesDiff.NodeDiff {
+		addedNodeList.AddNode(diffNodeAdded.Added)
+	}
+
+	return &addedNodeList
+}
+
+func (nlDiff *NodeListDiff) RemovedNodeList() *NodeList {
+	var removedNodeList NodeList
+	for _, removedNode := range nlDiff.NodesDiff.Removed {
+		removedNodeList.AddNode(removedNode)
+	}
+	for _, removedEdge := range nlDiff.EdgesDiff.Removed {
+		removedNodeList.AddEdge(removedEdge)
+	}
+	removedNodeList.RootElements = append(removedNodeList.RootElements, nlDiff.RootElmementsDiff.Removed...)
+
+	for _, diffNodeRemoved := range nlDiff.NodesDiff.NodeDiff {
+		removedNodeList.AddNode(diffNodeRemoved.Removed)
+	}
+
+	return &removedNodeList
+}
+
+// NodeListDiffNodes represents the differences between two NodeList nodes.
+type NodeListDiffNodes struct {
+	Added    []*Node
+	Removed  []*Node
+	NodeDiff []*NodeDiff
+}
+
+// NodeListDiffEdges represents the differences between two NodeList edges.
+type NodeListDiffEdges struct {
+	Added   []*Edge
+	Removed []*Edge
+}
+
+// NodeListRootElementDiff represents the differences between two NodeLists root elements.
+type NodeListRootElementDiff struct {
+	Added   []string
+	Removed []string
+}
+
+// Diff analyses a NodeList and returns a NodeList populated with all fields
+// that are different in nl2 from nl.
+func (nl *NodeList) Diff(nl2 *NodeList) NodeListDiff {
+	return NodeListDiff{
+		NodesDiff:         nl.diffNodes(nl2, false),
+		EdgesDiff:         nl.diffEdges(nl2),
+		RootElmementsDiff: nl.diffRootElements(nl2),
+	}
+}
+
+// DiffWithStripQulifiers analyses a NodeList and returns a NodeList populated with all fields
+// that are different in nl2 from nl after stripping identifier Quilifiers.
+func (nl *NodeList) DiffWithStripQulifiers(nl2 *NodeList, strip bool) NodeListDiff {
+	return NodeListDiff{
+		NodesDiff:         nl.diffNodes(nl2, strip),
```

Comment:
Wiring a stip parameter here will not scale well if we want to do other comparisons like this.

Instead, we should pass transformer functions that mutate the elements, I would suggest leaving the qualifier stripping for a follow-up pr to keep this PR easier to review.

Reply:
removed strip hack.
It was put in to be able to build a simple version drift command thought it basic enough a functionality to include out of the box as a option but i agree a more generic approach is needed.

Reply:
Cool, we will have methods for this but it is important to scope out the identifier matching logic from the diffin code.